### ISSUE TITLE
TfLiteStatus invoke_status = interpreter.Invoke(); Failed

### ISSUE THREAD

**Issue by aetherwind** [tensorflow/tensorflow#53293]
### 0. Summary

Help is appreciated!
Details are given below.

To do my first project with TF lite on microcontrollers (TFLM) I follow the Hello World example for TFLM.
I do this with my own network, though. The Structure is

- Input 1x16x16x1 
- Two CNN layers with maxpooling
- Flatten
- Dense layer
- Output 1x2

I encounter the following problem:

In the python world (nearly) everything works.
> Sidenote: It did not work to use converter.inference_input_type = tf.int8 (The quantization scaling seems to expect a uint8 input eventough int8 was specified) Therefore I use uint8.

Also in the Cpp world everything looks fine. Till I do interpreter.Invoke() This creates the follwoing error message in the terminal:
> Input UINT8, output INT8 not supported.
> Node QUANTIZE (number 0) failed to invoke with status 1

However, I still get a prediction that is incorrect.
I wanted to note that I use
converter.inference_input_type = tf.uint8
converter.inference_input_type = tf.uint8

Details follow:

### 1. System information

- Windows 10, use Cygwin to run the make process
- Packages (pip):
	numpy==1.19.2
	pandas==1.2.2
	tensorflow==2.4.0
	tflite==2.4.0
	tensorflow_hub==0.12.0
	opencv-python>=4.5.4
	tqdm==4.62.3
	matplotlib==3.4.3
	scikit-learn>=0.0.0
	pillow>=0.0.0
	scikit-image>=0.0.0
	tensorflow_model_optimization>=0.0.0
	ipykernel==6.4.1
	tensorboard>=0.0.0

### 2. Code

Provide code to help us reproduce your issues using one of the following options:

## Python-world (model produces reasonable results. Everything seems to be fine here.)
	# imports etc.

	# Load model
	model = tf.keras.models.load_model("src/mdl/mdl_KERAS_trained")

	# Define Representative Dataset
	data_obj = data_prep.gesture_data("../Data/lala")
	data_obj.load_data()
	data_obj.proc_data_picbased()
	data_obj.split_data()
	x_train = tf.convert_to_tensor(data_obj.x_train, dtype=tf.float32)
	y_train = tf.convert_to_tensor(data_obj.y_train, dtype=tf.float32)

	# Quick Check
	for i in range(12):
		print(model.predict(np.expand_dims(x_train[i, :, :, :], 0)))
		print(y_train[i, :])

	def representative_dataset_train():
		dataset = tf.data.Dataset.from_tensor_slices(x_train).batch(1)
		for data in dataset:
			yield [data]

	# Define converter
	converter = tf.lite.TFLiteConverter.from_keras_model(model)
	converter.optimizations = [tf.lite.Optimize.DEFAULT]
	converter.inference_input_type = tf.uint8
	converter.inference_output_type = tf.uint8
	converter.representative_dataset = representative_dataset_train
	converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
	converter.target_spec.supported_types = [tf.int8]

	# Convert
	tflite_model = converter.convert()

	# Save
	with open("src/mdl/mdl_tflite.tflite", "wb") as f:
		f.write(tflite_model)

	# Load TFLite
	interpreter = tf.lite.Interpreter(model_path="src/mdl/mdl_tflite.tflite")
	interpreter.allocate_tensors()

	# Get input and output tensors.
	input_details = interpreter.get_input_details()
	output_details = interpreter.get_output_details()

	# Scaling to UINT8
	input_quant = input_details[0]["quantization"]
	output_quant = output_details[0]["quantization"]

	x_train8 = np.fix((x_train) / input_quant[0] + input_quant[1])
	y_train8 = y_train

	# Quick check
	print("prediction")
	for i in range(20):
		input_data = np.array(
			np.expand_dims(x_train8[i, :, :, :], axis=0), dtype=np.uint8
		)
		interpreter.set_tensor(input_details[0]["index"], input_data)
		interpreter.invoke()
		output_data = interpreter.get_tensor(output_details[0]["index"])
		print(
			output_data, "   ",
			np.array(output_quant[0]) * (np.array(output_data) - np.array(output_quant[1])), "   ",
			y_train8[i],
		)

	
	
## Cpp-world: 
```

TF_LITE_MICRO_TESTS_BEGIN

TF_LITE_MICRO_TEST(LoadModelAndPerformInference) {
	// Define the input
	double x[16][16];	// Some input data which are numbers between -1. and 1. Import not shown for clarity

	// Set up logging
	tflite::MicroErrorReporter micro_error_reporter;

	// Model
	const tflite::Model* model = ::tflite::GetModel(g_mdl_model_data);
	if (model->version() != TFLITE_SCHEMA_VERSION) {
	  TF_LITE_REPORT_ERROR(&micro_error_reporter,
	                         "Model provided is schema version %d not equal "
	                         "to supported version %d.\n",
	                         model->version(), TFLITE_SCHEMA_VERSION);
	}

	// Resolver
	tflite::AllOpsResolver resolver;

	// Arena
	constexpr int kTensorArenaSize = 500*1024;
	uint8_t tensor_arena[kTensorArenaSize];

	// Interpreter
	tflite::MicroInterpreter interpreter(model, resolver, tensor_arena,
	                                       kTensorArenaSize, &micro_error_reporter);

	interpreter.AllocateTensors();

	// Input
	TfLiteTensor* input = interpreter.input(0);

	// Does Input match
	TF_LITE_MICRO_EXPECT_NE(nullptr, input);
	TF_LITE_MICRO_EXPECT_EQ(4, input->dims->size);
	TF_LITE_MICRO_EXPECT_EQ( 1, input->dims->data[0]);
	TF_LITE_MICRO_EXPECT_EQ(16, input->dims->data[1]);
	TF_LITE_MICRO_EXPECT_EQ(16, input->dims->data[2]);
	TF_LITE_MICRO_EXPECT_EQ( 1, input->dims->data[3]);
	TF_LITE_MICRO_EXPECT_EQ(kTfLiteUInt8, input->type);		//kTfLiteUInt8


	// Get quantization parameters
	double input_scale = input->params.scale;
	int input_zero_point = input->params.zero_point;


	// Quantize
	uint8_t x_quantized[256];
	for(int i=0;i<ni;i++){
		for(int j=0;j<nj;j++){
			x_quantized[i*nj+j] = (uint8_t) ((x[i][j]/input_scale) + input_zero_point);
			input->data.uint8[i*nj+j] = x_quantized[i*nj+j];
		}
	}


	// Run the model
	MicroPrintf("-----invoke-----");
	TfLiteStatus invoke_status = interpreter.Invoke();
	MicroPrintf("-----check-----");
	TF_LITE_MICRO_EXPECT_EQ(kTfLiteOk, invoke_status);

	MicroPrintf("End of File");


}

TF_LITE_MICRO_TESTS_END
```

### 3. Failure after conversion

Output of make process:
```

	$ make --no-silent --trace -f tensorflow/lite/micro/tools/make/Makefile test_mdl_test
	
	...

	[Some lines of output]
	...
	
	tensorflow/lite/micro/tools/make/gen/windows_x86_64_default/bin/mdl_test '~~~ALL TESTS PASSED~~~' windows
	Testing LoadModelAndPerformInference
	----------
	input_scale 1.0039215*2^-7
	input_zero_point 127
	----------
	-----invoke-----
	Input UINT8, output INT8 not supported.
	Node QUANTIZE (number 0) failed to invoke with status 1
	-----check-----
	kTfLiteOk == invoke_status failed at tensorflow/lite/micro/examples/mdl/mdl_test.cc:200 (0 vs 1)
	output_scale 1.0*2^-8
	output_zero_point 0
	----------
	prediction:   3   4
	prediction:   3   4
	End of File
	0/1 tests passed
	~~~SOME TESTS FAILED~~~
	~~~SOME TESTS FAILED~~~

	make: *** [tensorflow/lite/micro/examples/mdl/Makefile.inc:40: test_mdl_test] Error 1

```

**Comment by sushreebarsa** [comment]
@aetherwind Could you please try with the latest TF v2.7.0 and let us know the outcome ? Thanks!

**Comment by aetherwind** [comment]
@sushreebarsa Thanks for helping. 

#### I went to TF 2.7 However the result stays the same. 
pip list:
> tensorboard                  2.7.0
> tensorboard-data-server      0.6.1
> tensorboard-plugin-wit       1.8.0
> tensorflow                   2.7.0
> tensorflow-estimator         2.7.0
> tensorflow-io-gcs-filesystem 0.22.0



#### What is in my opinion interesting is the following: I run the checks:
```
TF_LITE_MICRO_EXPECT_EQ(kTfLiteUInt8, input->type);
TF_LITE_MICRO_EXPECT_EQ(kTfLiteUInt8, ouptut->type);
```
They do not raise an error.

However I get an output that says 'Input UINT8, output INT8 not supported.' (I mean, where does the latter come from? Even in the C world the model clearly says that it has output UINT8 when running the check on output)



#### Also I get the following warnings during conversion in Python:
2021-12-06 09:08:47.600087: W tensorflow/python/util/util.cc:368] Sets are not currently considered sequences, but this may change in the future, so consider avoiding using them.
2021-12-06 09:08:49.281014: W tensorflow/compiler/mlir/lite/python/tf_tfl_flatbuffer_helpers.cc:363] Ignored output_format.
2021-12-06 09:08:49.282491: W tensorflow/compiler/mlir/lite/python/tf_tfl_flatbuffer_helpers.cc:366] Ignored drop_control_dependency.

However those warnings also show up in https://www.tensorflow.org/lite/guide/signatures. Therefore, I considered them 'normal'.

Do you have any ideas what the problem could be?

**Comment by sachinprasadhs** [comment]
Did you follow the step mentioned [here](https://www.tensorflow.org/lite/microcontrollers/get_started_low_level#7_allocate_memory). Thanks!

**Comment by aetherwind** [comment]
Hello sachinprasadhs,

thanks for your reply. To be honest I simply went for a large memory size for testing, like 500*1024, assuming this would cover all my needs and worry about getting it smaller at a latter point. 

I now tried and it works wit 5*1024.

Do I have to worry? As This is approx. 5000 bytes, but my model has approx. 17000 parameters. Somehow this does not seem to match.

**Comment by aetherwind** [comment]
The error seems to be invoked in micro_graph.cc at line 183

```
    if (invoke_status == kTfLiteError) {
      MicroPrintf("Node %s (number %d) failed to invoke with status %d",
                  OpNameFromRegistration(registration), i, invoke_status);
      return kTfLiteError;
    } else if (invoke_status != kTfLiteOk) {
      return invoke_status;
    }
```

**Comment by aetherwind** [comment]
When using 
```
converter.inference_input_type = tf.int8
converter.inference_output_type = tf.int8
```
instead of
```
converter.inference_input_type = tf.uint8
converter.inference_output_type = tf.uint8
```
the tests do not invoke errors anymore. However, the results of the quantized model are wrong - already in the python world.

**Comment by sachinprasadhs** [comment]
If your issue is resolved could you please move this to closed. Thanks!

**Comment by aetherwind** [comment]
It is not resolved. The core issue still remains:  The uint8 quantisation is not working in cpp for me and i cannot figure out at which point it goes wrong.

**Comment by aetherwind** [comment]
When running the original Hello World after changing input and output quantization from int8 to uint8 I get the following error when making the Cpp (running with int8 quantization works):

```
$ make -f tensorflow/lite/micro/tools/make/Makefile test_hello_world_test
find: ‘../google/’: No such file or directory
tensorflow/lite/micro/tools/make/downloads/flatbuffers already exists, skipping the download.
tensorflow/lite/micro/tools/make/downloads/kissfft already exists, skipping the download.
tensorflow/lite/micro/tools/make/downloads/pigweed already exists, skipping the download.
g++ -std=c++11 -fno-rtti -fno-exceptions -fno-threadsafe-statics -Werror -fno-unwind-tables -ffunction-sections -fdata-sections -fmessage-length=0 -DTF_LITE_STATIC_MEMORY -DTF_LITE_DISABLE_X86_NEON -Wsign-compare -Wdouble-promotion -Wshadow -Wunused-variable -Wmissing-field-initializers -Wunused-function -Wswitch -Wvla -Wall -Wextra -Wstrict-aliasing -Wno-unused-parameter  -DTF_LITE_USE_CTIME -Os -I. -Itensorflow/lite/micro/tools/make/downloads/gemmlowp -Itensorflow/lite/micro/tools/make/downloads/flatbuffers/include -Itensorflow/lite/micro/tools/make/downloads/ruy -Itensorflow/lite/micro/tools/make/gen/windows_x86_64_default/genfiles/ -Itensorflow/lite/micro/tools/make/downloads/kissfft -c tensorflow/lite/micro/tools/make/gen/windows_x86_64_default/genfiles/tensorflow/lite/micro/examples/hello_world/hello_world_model_data.cc -o tensorflow/lite/micro/tools/make/gen/windows_x86_64_default/obj/core/tensorflow/lite/micro/tools/make/gen/windows_x86_64_default/genfiles/tensorflow/lite/micro/examples/hello_world/hello_world_model_data.o
g++ -std=c++11 -fno-rtti -fno-exceptions -fno-threadsafe-statics -Werror -fno-unwind-tables -ffunction-sections -fdata-sections -fmessage-length=0 -DTF_LITE_STATIC_MEMORY -DTF_LITE_DISABLE_X86_NEON -Wsign-compare -Wdouble-promotion -Wshadow -Wunused-variable -Wmissing-field-initializers -Wunused-function -Wswitch -Wvla -Wall -Wextra -Wstrict-aliasing -Wno-unused-parameter  -DTF_LITE_USE_CTIME -I. -Itensorflow/lite/micro/tools/make/downloads/gemmlowp -Itensorflow/lite/micro/tools/make/downloads/flatbuffers/include -Itensorflow/lite/micro/tools/make/downloads/ruy -Itensorflow/lite/micro/tools/make/gen/windows_x86_64_default/genfiles/ -Itensorflow/lite/micro/tools/make/downloads/kissfft -o tensorflow/lite/micro/tools/make/gen/windows_x86_64_default/bin/hello_world_test tensorflow/lite/micro/tools/make/gen/windows_x86_64_default/obj/core/tensorflow/lite/micro/examples/hello_world/hello_world_test.o tensorflow/lite/micro/tools/make/gen/windows_x86_64_default/obj/core/tensorflow/lite/micro/tools/make/gen/windows_x86_64_default/genfiles/tensorflow/lite/micro/examples/hello_world/hello_world_model_data.o tensorflow/lite/micro/tools/make/gen/windows_x86_64_default/lib/libtensorflow-microlite.a -Wl,--fatal-warnings -Wl,--gc-sections -lm
tensorflow/lite/micro/tools/make/gen/windows_x86_64_default/bin/hello_world_test '~~~ALL TESTS PASSED~~~' windows
Testing LoadModelAndPerformInference
tensorflow/lite/micro/kernels/quantize_common.cc:54 input->type == kTfLiteFloat32 || input->type == kTfLiteInt32 || input->type == kTfLiteInt16 || input-> was not true.
Node QUANTIZE (number 0f) failed to prepare with status 1
interpreter.AllocateTensors() == kTfLiteOk failed at tensorflow/lite/micro/examples/hello_world/hello_world_test.cc:58 (1 vs 0)
make: *** [tensorflow/lite/micro/examples/hello_world/Makefile.inc:38: test_hello_world_test] Segmentation fault (core dumped)
```

Is uint8 currently really an implemented quantization?

**Comment by sachinprasadhs** [comment]
You can refer [this](https://www.tensorflow.org/lite/performance/quantization_spec) document for details.

**Comment by aetherwind** [comment]
Hello sachinprasadhs,
This information is really helpful. Thank you very much!! :)

> Note: In the past our quantization tooling used per-tensor, asymmetric, uint8 quantization. New tooling, reference kernels, and optimized kernels for 8-bit quantization will use this spec [(int8)].

I did not really find this information, especially, there was no need looking for it, as [https://www.tensorflow.org/lite/performance/post_training_integer_quant](url) (which is the same website) talks about uint8 as it would be the thing to do, e.g. this code example:
```
# Set the input and output tensors to uint8 (APIs added in r2.3)
converter.inference_input_type = tf.uint8
converter.inference_output_type = tf.uint8
```
But seems the feature was removed again.

Anyway! Thanks very much for your support!!
